i#2717 Change the implementation of options.

nilayvaish opened the pull request:
Curently the functions for setting and getting options have code for
each option generated using preprocessor macros.  This results in
excessive amount of code and increases the size of the library.  With
this commit, we alter the functions for setting and getting options.

* We define a new data type struct option_t.  This struct contains all
the information that we need to process an option.  Using macros, we
have created an array that contains the required information for each
option.

* We also define certain helper types and functions.

* The code for getting and setting options now contains a for loop in
which we go over all the options one by one.  This reduces the amount
of code by a significant amount.

This reduces the size of the static library by about 340 KB.

Issue: #2717

derekbruening commented on core/options.c:
```diff
@@ -2438,7 +2724,7 @@ show_dynamo_option_descriptions()
     }
 #include "optionsx.h"
-#undef OPTION_COMMAND
+# undef OPTION_COMMAND
```
One or the other's indentation is off: if there's a containing ifdef both should have a space after `#` (or multiple if inside nested #ifs) else no space.

nilayvaish replied:
Fixed.

derekbruening commented on core/options.c:
```diff
@@ -104,6 +104,64 @@ static void ignore_varargs_function(char *format, ...) { }
+ * since never written.
+ */
+static const option_t internal_option_traits[] = {
+#  include "optionsx.h"
```
I know some of these were like this before but while we're changing these lines we may as well match the current style: single space after `#` since inside single level of #if, here and defines above.

nilayvaish replied:
Fixed.

derekbruening commented:
Note the WIndows build failure as well.

fhahn commented:
run aarch64 tests

fhahn commented:
run aarch64 tests

nilayvaish commented on core/options.c:
```diff
@@ -104,6 +104,64 @@ static void ignore_varargs_function(char *format, ...) { }
 
 #endif /* NOT_DYNAMORIO_CORE */
 
+#include <stddef.h>
```
Done.

nilayvaish commented on core/options.c:
```diff
@@ -708,20 +1003,17 @@ get_pcache_dynamo_options_string(options_t *options, char *opstr, int len,
+    int i;
+    for (i = 0; i < num_options; ++i) {
+        if (option_traits[i].affects_pcache >= pcache_effect &&
+            option_traits[i].name[0] != ' ') { /* not synthethic */
```
Fixed.

nilayvaish commented on core/options.c:
```diff
@@ -708,20 +1003,17 @@ get_pcache_dynamo_options_string(options_t *options, char *opstr, int len,
-                       description, flag, pcache) {                     \
-        if (command_line_option[0] != ' ' && /* not synthethic */       \
-            (op_pcache_t) OPTION_AFFECTS_PCACHE_##name >= pcache_effect && \
-            DIFF_##type(options->name,default_options.name)) {          \
```
Fixed.

nilayvaish commented on core/options.c:
```diff
@@ -677,21 +974,19 @@ get_dynamo_options_string(options_t *options, char *opstr, int len, bool minimal
-                       description, flag, pcache) {                     \
-        if (command_line_option[0] != ' ' && /* not synthethic */       \
-            (!minimal ||                                                \
-             DIFF_##type(options->name,default_options.name))) {        \
```
Fixed.

nilayvaish commented on core/options.c:
```diff
@@ -104,6 +104,64 @@ static void ignore_varargs_function(char *format, ...) { }
+} option_type_t;
+
+/* Function for identifying string type. */
+bool
```
Done.

nilayvaish commented on core/options.c:
```diff
@@ -104,6 +104,64 @@ static void ignore_varargs_function(char *format, ...) { }
+#  define OPTION_COMMAND(type, name, default_value, command_line_option, \
+                         statement, description, flag, pcache) /* nothing */
+
+/* Read only source for default internal option values and names no lock needed
```
Done.

nilayvaish commented on core/options.c:
```diff
@@ -485,28 +556,75 @@ parse_liststring_t(liststring_t *var, void *value) {
+/* We mark this function to be NOINLINE so that in case the compiler unrolls the
+ * loop where this function is used, this function is not copied over there
+ * mutliple times.  Copying over this code increases code size significantly
+ * especially since strcmp() is a declared either as macro or as inline.
```
Done.

nilayvaish commented on core/options.c:
```diff
@@ -485,28 +556,75 @@ parse_liststring_t(liststring_t *var, void *value) {
+{
+    if (strcmp(opt+1, command_line_option) == 0) {
+        *value = getword(optstr, pos, wordbuffer, max_option_length);
+          /* FIXME: check argument */
```
Fixed indentation.

nilayvaish commented on core/options.c:
```diff
@@ -533,20 +650,28 @@ set_dynamo_options_common(options_t *options, const char *optstr, bool for_this_
+                                    wordbuffer, sizeof(wordbuffer), &value);
+                }
+                if (value != NULL) {
+                    void* optptr = (char*)(options) + option_traits[i].offset;
```
Fixed.

nilayvaish commented on core/options.c:
```diff
@@ -677,21 +974,19 @@ get_dynamo_options_string(options_t *options, char *opstr, int len, bool minimal
+    if (!minimal) {
+        int i;
+        for (i = 0; i < num_options; ++i) {
+            if (option_traits[i].name[0] != ' ') { /* not synthethic */
```
Fixed spelling.

nilayvaish commented on core/options.c:
```diff
@@ -677,21 +974,19 @@ get_dynamo_options_string(options_t *options, char *opstr, int len, bool minimal
+                const void *val_ptr =
+                    (char*)(options) + option_traits[i].offset;
+                print_option_type(option_traits[i].type, optionbuff, val_ptr,
+                         option_traits[i].name);
```
Fixed indentation.

nilayvaish commented:
Hi Derek

Here are the changes that went into the update patch:
* Fixed the error related to compilation on Windows.  I was using uint64 for offset and size of members of struct options_t and then using those for computing the address to use for accessing particular options.  That was causing problem in 32-bit mode as compiler reasoned that pointer arithmetic would be incorrect.  I have changed that uint64 to uint and things are working fine for Windows build.
* Fixed the couple of missing calls to diff_by_type() function you pointed out in your review.
* Fixed the style errors.

Please take another look.

Thanks
Nilay

derekbruening commented:
Are you rewriting history in your branch?  When I try to see just the changes since I last reviewed there's just the define move and what is labeled as the original commit yet it has changes from the review.

nilayvaish commented:
Yes, I am rewriting the history.  Now that I think about it, it would make the reviewer's job difficult since they would not be able to make out what changed.  From now on, I would make changes and commit them as new commits (instead of amending existing commits).  As a final step, I will rewrite the history so as to present a clean picture.

derekbruening commented:
No, please do not rewrite any history of git shared branches.  Please see the code review and workflow wiki pages: the final squash-and-merge into master cleans everything up into a single commit on master.  The feature branch should use separate commits, and just like for git in general, not just our project, rewriting history on a shared branch is pretty much always to be avoided.

derekbruening commented:
run aarch64 tests

derekbruening commented:
Something is not right with this change.  Compare the printout of non-default options before this went in:

```
<Initial options = -no_dynamic_options -code_api -stack_size 56K -signal_stack_size 32K -max_elide_jmp 0 -max_elide_call 0 -early_inject -emulate_brk -no_inline_ignored_syscalls -native_exec_default_list '' -no_native_exec_managed_code -no_indcall2direct >
```

And now after this is committed:
```
<Initial options = -no_dynamic_options -code_api -stack_size 56K -signal_stack_size 32K -max_elide_jmp 0 -max_elide_call 0 -vm_base 0x0000000046000000 -vm_max_offset 0x0000000010000000 -early_inject -early_inject_address 0x0000000000000000 -emulate_brk -no_inline_ignored_syscalls -native_exec_default_list '' -no_native_exec_managed_code -no_indcall2direct -stress_fake_userva 0x0000000000000000 >
```

derekbruening commented:
@nilayvaish I assume you're looking into this

derekbruening commented:
This is also hitting build errors on AArch64 in CDash, despite the precommit here succeeding, but it's not clear why:
http://dynamorio.org/CDash/buildSummary.php?buildid=32063
```
*Error* (first 5)
CMake Error at /.../src/core/CMake_readelf.cmake:94 (message):
  *** Error:
  /.../run/build_release-external-64/lib64/release/libdynamorio.so
  has undefined symbol: GLOBAL DEFAULT UND __strncat_chk
```

nilayvaish commented:
Yes.  I am looking into it.

On 13 December 2017 at 21:46, Derek Bruening <notifications@github.com>
wrote:

> @nilayvaish <https://github.com/nilayvaish> I assume you're looking into
> this
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/DynamoRIO/dynamorio/pull/2748#issuecomment-351615124>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/APgHiXv_NGO8H1ua1dC0RAFodZ2eRW_tks5tALZMgaJpZM4Q7Y3q>
> .
>

derekbruening commented:
One more issue is this config:

```
$ cmake -DVMAP=OFF -DVPS=ON ../src
￼```

Something changed regarding internal vs external options or maybe CLIENT_INTERFACE options being exposed vs being constants:

http://dynamorio.org/CDash/buildSummary.php?buildid=32089

```
*Error* (first 5)
core/options.c line 143 (http://dynamorio.org/CDash/viewBuildError.php?type=0&buildid=32089)
/work/dr/nightly/src/core/options.c:143:24: error: ‘options_t {aka struct _options_t}’ has no member named ‘nolink’
core/options.c line 143 (http://dynamorio.org/CDash/viewBuildError.php?type=0&buildid=32089)
/work/dr/nightly/src/core/options.c:143:24: error: ‘options_t {aka struct _options_t}’ has no member named ‘link_ibl’
core/options.c line 143 (http://dynamorio.org/CDash/viewBuildError.php?type=0&buildid=32089)
/work/dr/nightly/src/core/options.c:143:24: error: ‘options_t {aka struct _options_t}’ has no member named ‘tracedump_binary’
core/options.c line 143 (http://dynamorio.org/CDash/viewBuildError.php?type=0&buildid=32089)
/work/dr/nightly/src/core/options.c:143:24: error: ‘options_t {aka struct _options_t}’ has no member named ‘tracedump_text’
core/options.c line 143 (http://dynamorio.org/CDash/viewBuildError.php?type=0&buildid=32089)
/work/dr/nightly/src/core/options.c:143:24: error: ‘options_t {aka struct _options_t}’ has no member named ‘tracedump_origins’
```

derekbruening commented:
(For the printout of non-default options above, that's printed by default just running debug build on any app, `bin64/drrun -- ls`)

derekbruening commented:
This is now making AArch64 precommit fail.  Unless there's an imminent fix maybe we should revert?

nilayvaish commented:
Just an update.  We had identified three problems with this commit.  Two of those, different print output for  options and compilation failure of of internal options have been fixed.  The third one, linking failure on AArch64, has been reproduced with a VM.  The actual reason why the symbol __strncat_chk is added is still unclear.
